### PR TITLE
Validate bundle list when creating or updating a profile

### PR DESCRIPTION
**Pull request by Rohitrajak1807**
- Changed bundles flag to accept a slice of strings
- Add regex validation for bundle list when creating or updating a profile
- Add regex validation when creating, dump or updating a bundle
- When updating and creating profiles we check of the supplied bundles exist or not

### Manual:
```
➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon bundle create xenial-static-f --tags app --desc xenial --from-file ./examples/bundles/xenial.yaml

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon bundle create xenial-static-f- --tags app --desc xenial --from-file ./examples/bundles/xenial.yaml
Error: invalid resource name: xenial-static-f-

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon bundle create xenial-static-f-this-is-a-very-very-very-very-long-name-which-is-not-allowed-and-i-am-sure-that-this-command-will-fail --tags app --desc xenial --from-file ./examples/bundles/xenial.yaml
Error: invalid resource name: xenial-static-f-this-is-a-very-very-very-very-long-name-which-is-not-allowed-and-i-am-sure-that-this-command-will-fail

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon bundle update xenial-static-f- --from-file ./examples/bundles/guestbook.yaml
Error: invalid resource name: xenial-static-f-

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon bundle update xenial-static-f --from-file ./examples/bundles/guestbook.yaml

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon profile list
NAME       GEN  TYPE     BUNDLES                           REPO-URL                                        REPO-PATH           OVRDS  TAGS        DESCRIPTION
static-1   2    static   [guestbook-static xenial-static]  (N/A)                                           (N/A)               0      [examples]  static profile 1
dynamic-1  2    dynamic  [xenial-static]                   https://github.com/jayanth-tjvrr/pf9arlon-test  profiles/dynamic-1  0      [examples]  dynamic test 1

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon profile create static-f --bundles xenial-static-f,guestbook-static,xenial-static --desc "testing" --tags "example"

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon profile create static-f_ --bundles xenial-static-f,guestbook-static,xenial-static --desc "testing" --tags "example"
Error: failed to create profile configmap: Profile.core.arlon.io "static-f_" is invalid: metadata.name: Invalid value: "static-f_": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon profile create static-f --bundles xenial-static-f_,guestbook-static,xenial-static --desc "testing" --tags "example"
Error: invalid resource name: xenial-static-f_

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon profile create dynamic-f --repo-url "https://github.com/jayanth-tjvrr/pf9arlon-test" --repo-base-path profiles --bundles guestbook-static,xenial-static,xenial-static-f --desc "test" --tags "test"

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon profile create dynamic-f[ --repo-url "https://github.com/jayanth-tjvrr/pf9arlon-test" --repo-base-path profiles --bundles guestbook-static,xenial-static,xenial-static-f --desc "test" --tags "test"¯
zsh: bad pattern: dynamic-f[

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon profile update dynamic-f --bundles xenial-static-f --desc "test" --tags "test"¯

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon profile update dynamic-f^ --bundles xenial-static-f,guestbook-static
zsh: no matches found: dynamic-f^

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon profile update 'dynamic-f^' --bundles xenial-static-f,guestbook-static
Error: failed to get augmented profile: failed to get profile: failed to get profile from configmap: failed to get profile configmap: configmaps "dynamic-f^" not found

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon profile update dynamic-f --bundles xenial-static-f:,guestbook-static
Error: invalid resource name: xenial-static-f:

➜  arlon git:(private/Rohitrajak1807/validate-bundle) bin/arlon profile update dynamic-f --bundles xenial-static-f:,guestbook-static,404-not-found
Error: invalid resource name: xenial-static-f:

➜  arlon git:(private/Rohitrajak1807/validate-bundle) ✗ bin/arlon profile create dynamic-ff --repo-url "https://github.com/jayanth-tjvrr/pf9arlon-test" --repo-base-path profiles --bundles 404-not-found,guestbook-static,xenial-static,xenial-static-f --desc "test" --tags "test"
Error: one or more supplied bundles don't exist

➜  arlon git:(private/Rohitrajak1807/validate-bundle) ✗ bin/arlon profile update dynamic-f --bundles xenial-static-f,guestbook-static,404-not-found
Error: one or more supplied bundles don't exist
```

### CODE REVIEW

File: cmd/bundle/list.go [hunk 1]
```diff
@@ -36,39 +34,20 @@ func listBundlesCommand() *cobra.Command {
...
-	}
-	secrets, err := secretsApi.List(context.Background(), opts)
+	bundles, err := bundle.List(config, ns)
 	if err != nil {
-		return fmt.Errorf("failed to list secrets: %s", err)
-	}
-	if len(secrets.Items) == 0 {
-		fmt.Println("no bundles found")
-		return nil
+		if errors.Is(err, bundle.ErrNoBundles) {
```
Comment (bcle) [inline cmd/bundle/list.go:39]: Absence of bundles should not be an error. See related comment in the function.

File: cmd/profile/create.go [hunk 2]
```diff
@@ -16,7 +16,7 @@ func createProfileCommand() *cobra.Command {
 	var arlonNs string
 	var argocdNs string
 	var desc string
-	var bundles string
+	var bundles = make([]string, 1)
```
Comment (bcle) [inline cmd/profile/create.go:19]: Is it necessary to make() the variable if it's going to be reset to nil by StringSliceVar()?

File: pkg/bundle/list.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	secretsApi := corev1.Secrets(namespace)
+	opts := metav1.ListOptions{
+		LabelSelector: "managed-by=arlon,arlon-type=config-bundle",
+	}
+	secrets, err := secretsApi.List(context.Background(), opts)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrFailedList, err)
+	}
+	if len(secrets.Items) == 0 {
+		return nil, ErrNoBundles
```
Comment (bcle) [inline pkg/bundle/list.go:41]: The absence of bundles should not be treated as error. In general, this applies to all List operations.

**Review by bcle — CHANGES_REQUESTED** [review]

File: cmd/profile/create.go (hunk 2 above)
Comment (Rohitrajak1807) [reply cmd/profile/create.go:19]: will test and post the results.

File: pkg/profile/update.go [hunk 4]
```diff
@@ -18,11 +19,24 @@ func Update(
...
 	overrides []arlonv1.Override,
 ) (dirty bool, err error) {
+	for _, name := range bundlesPtr {
+		if !bundle.IsValidK8sName(name) {
+			return false, fmt.Errorf("%w: %s", bundle.ErrInvalidName, name)
+		}
+	}
+	existingBundles, err := bundle.List(config, arlonNs)
+	if err != nil {
+		return false, err
```
Comment (bcle) [inline pkg/profile/update.go:34]: To build a cumulative string-based "stack trace" to help the caller identify the chain of calls leading to the error, don't just return the error unmodified, but instead return something like `fmt.Errorf("failed to list bundles: %s", err)`

**Review by bcle — CHANGES_REQUESTED** [review]
Can you fix the one issue about returning an error. The rest looks ok.

File: pkg/bundle/list.go (hunk 3 above)
Comment (Rohitrajak1807) [reply pkg/bundle/list.go]: done

File: cmd/profile/create.go (hunk 2 above)
Comment (Rohitrajak1807) [reply cmd/profile/create.go]: works

Comment (Rohitrajak1807) [reply cmd/profile/create.go]: works! the change is in

File: pkg/bundle/list.go (hunk 3 above)
Comment (Rohitrajak1807) [reply pkg/bundle/list.go]: done

File: pkg/profile/update.go (hunk 4 above)
Comment (Rohitrajak1807) [reply pkg/profile/update.go:34]: This is in profile/update, so could I add a message saying: `failed to get existing bundles` ?

File: cmd/bundle/list.go (hunk 1 above)
Comment (Rohitrajak1807) [reply cmd/bundle/list.go]: done

File: pkg/profile/update.go (hunk 4 above)
Comment (Rohitrajak1807) [reply pkg/profile/update.go:34]:
```
Error: failed to list bundles: failed to list secrets: Get "https://pf9-arlon.platform9.horse/qbert/v4/e5d5d7327829479ab04b050bf52a946e/sunpike/api/v1/namespaces/arlon/secrets?labelSelector=managed-by%3Darlon%2Carlon-type%3Dconfig-bundle": dial tcp: lookup pf9-arlon.platform9.horse: no such host
```

The error looks tike this now.

**Review by bcle — APPROVED** [review]

File: pkg/profile/update.go [hunk 5]
```diff
@@ -43,7 +57,7 @@ func Update(
 	if bundlesPtr == nil {
 		bundles = CommaSeparatedFromStringList(prof.Spec.Bundles)
 	} else {
-		bundles = *bundlesPtr
+		bundles = CommaSeparatedFromStringList(bundlesPtr)
```
Comment (cruizen) [inline pkg/profile/update.go:60]: I find the name `CommaSeparatedFromStringList` hard to parse. Does `GetConcatenatedStringFromBundlesList` or `ConcatenateBundlesFromList` make more sense?

**Comment by Rohitrajak1807** [comment]
merging
